README.md
# dynamicform
![pub](https://img.shields.io/badge/pub-v0.23.0-orange) ![GitHub](https://img.shields.io/github/license/liodali/checkbox_grouped)


create your form with easier way

## Getting Started

 * Generate custom form
 * Form Controller to manage form
 * login Form
 * payment form
 * Pre-existing elements

## Installing

Add the following to your `pubspec.yaml` file:

    dependencies:
		dynamic_form: ^0.23.0




## Simple Usage
#### Creating a basic `SimpleDynamicForm`:

```dart
        SimpleDynamicForm(
                  controller: controller,
                  groupElements: [
                          GroupElement(
                                        directionGroup: DirectionGroup.Vertical,
                                        textElements: [
                                        TextElement(
                                          id:"name",
                                          label: "name",
                                        ),
                                        TextElement(
                                            id:"password",
                                             label: "password", 
                                             typeInput: TypeInput.Password,
                                        )
                                    ],
                                )
                              ],
                          );
```



####  `Properties in SimpleDynamicForm`

 | Properties                     | Description                                                    |
 | ------------------------------ | -------------------------------------------------------------- |
 | `groupElements`                | list of element to build your form                             |
 | `padding`                      |  The amount of space by which to inset the form                |
 | `controller`                   |  The FormController to get values and validate your form       |
 | `submitButton`                 |  (Widget) custom submit Widget that you want to add to the form |


### Declare FormController to get validation,list values of forms

```dart
 FormController controller = FormController();
```
> you can access to controller from your submit button
>
```dart
 FormController controller = SimpleDynamicForm.of(context);
```
### validate forms

```dart
controller.validate();
```

### show error to forms fields  after validation
* error will be cleared automatically when form validate
```dart
 controller.addErrorToField(String idField,String errorMessage);
```


### recuperate all values in form :
```dart
 controller.getAllValues();
```

> if you are used ids in element,  you can recuperate values with their ids

```dart
 Map values = controller.getByIds();
```

> you recuperate by id

```dart
 controller.getValueById(id);
```
### set input value in form :

> you can change field value using id of that field

```dart
 controller.setFieldValueById(id,value);
```
### clear  all inputs in form :

```dart
 controller.clearValues();
```


## LoginForm
> pre-existing login form to make easy for you to build

### Simple Usage

#### Creating a basic `LoginForm` :


```dart
  
    LoginForm(
          controller:controller
          submitLogin:RaisedButton(
           onPressed: () {
                final email = controller.email;
                final password = controller.password;
                print("$email,$password");
              },
          child: Text("Log In"),
         ),
          onlyEmail: false,
          textButton: Text("Log IN"),
          paddingFields: const EdgeInsets.all(0),
          decorationLoginForm: DecorationLoginForm(
            commonDecoration: decoration,
            decorationEmailElement: OutlineDecorationElement(
                contentPadding: EdgeInsets.only(
                left: 12.0,
              ),
            ),
            decorationPasswordElement: OutlinePasswordElementDecoration(
                contentPadding: EdgeInsets.only(left: 8),
                prefix: Padding(
                   padding: EdgeInsets.all(6.0),
                   child: Icon(
                      Icons.lock,
                      size: 20,
                  ),
            ),
            enableVisibilityPassword: false
      ),
    )
```


### LoginFormController 

 * validate login form
 * recuperate email/password
 * show error in specific input

1) initialization

```dart
LoginFormController controller = LoginFormController();
```
2) recuperate email/Username,password
```dart
final email = controller.email;
final password = controller.password;
```
3) access to controller in submit button 

```dart
 LoginFormController controller = LoginForm.of(context)
```


4) show field error after validation (use case when auth failed)

```dart
 controller.addEmailError("invalid Email not found");
 controller.addPasswordError("invalid Email not found");
```


5) Properties

 | Properties                     | Description                                                    |
 | ------------------------------ | -------------------------------------------------------------- |
 | `controller`                   | LoginFormController to validate login form and get data                             |
 | `decorationLoginForm`          | (DecorationLoginForm) contain configuration of the inputs decoration of email/password fields in form                             |
 | `directionGroup`               | Direction of form (Vertical/Horizontal)                        |
 | `paddingFields`                | padding between fields                                         |
 | `onlyEmail`                    | enable only email type textField                               |
 | `passwordError`                | messages errors to show  when password field not validate      |
 | `usernameEmailError`           | messages errors to show when email/username not validate       |
 | `submitLogin`                  | (Widget) Button of submit form                |


## PaymentForm

* pre-existing payment form to make easy for you to build

1) Simple Usage

#### Creating a basic `PaymentForm` :

```dart
    PaymentForm(
          controller:controller,
          decorationElement: OutlineDecorationElement(),
          errorMessageCVV: "cvv is invalid",
          errorMessageDateExpiration: "date expiration is invalid",
          errorIsRequiredMessage: "This field  is required",
          labelCVV: "cvv",
          labelDateExpiration: "date expiration",
          labelCardNumber: "card number",
          submitButton: ElevatedButton(
            onPressed: () {
              controller.validate();
            },
            child: Text("pay"),
          ),
        )
```

### PaymentController

1) initialization

```dart
PaymentController controller = PaymentController();
```
2) validation payment form
```dart
bool isValid = controller.validate();
```
3) recuperate cardNumber,cvv,dateExpiration

```dart
final cardNumber = controller.cardNumber;
final cvv = controller.cvv;
final dateExpiration = controller.dateExpiration;
```
5) access to payment controller inside payment submit button

```dart
 PaymentController controller = PaymentForm.of(context);
```

6) show field error after validation (use case when card check failed)

```dart
 controller.addCardNumberError(errorMessage);
 controller.addCVVError(errorMessage);
 controller.addDateExpirationError(errorMessage);
```

7) `Properties

 | Properties                     | Description                                                    |
 | ------------------------------ | -------------------------------------------------------------- |
 | `controller`                   | (PaymentController) controller to validate form,setError fields,clear values                           |
 | `entryModeDateExpiration`      | (DateExpirationEntryMode)   input type of card date expiration can be dropdown or input(textField)                           |
 | `decorationElement`            | decoration of all input field in form                             |
 | `buttonDecoration`             | decoration of button that contain radius,backgroundColor,width |
 | `errorMessageDateExpiration`   | messages errors to show  when Date Expiration field not validate      |
 | `errorMessageCVV`              | messages errors to show when cvv field is invalidate            |
 | `errorMessageCardNumber`       | messages errors to show when credit card number is invalidate   |
 | `errorIsRequiredMessage`       | messages errors to show when at least one field not filled      |
 | `labelCardNumber`              | text label of credit card number field                          |
 | `labelDateExpiration`          | text label of date expiration field                             |
 | `labelCVV`                     | text label of cvv field                                         |
 | `submitButton`                 | (widget) submit button widget                                         |


## API
### How to sketch your form ?

> ` textElement is small element in dynamicForm `
> ` GroupElement is group of TextElement`

####  `GroupElement`
| Properties          | Description                                                                                  |
| ------------------- | -------------------------------------------------------------------------------------------- |
| `directionGroup`    | Direction of form (Vertical/Horizontal)                                                      |
| `sizeElements`      | size of each textElement  of form When direction Horizontal,sum of values should be egal a 1 |
| `textElements`      | group of textElement.                                                                        |
| `padding`           | padding of groups.                                                                           |
| `decoration`        | decoration  of container groups.                                                             |
| `backgroundColor`   | color of the container groups.                                                               |


####  `textElement`
| Properties          | Description                                    |
| ------------------- | ---------------------------------------------- |
| `typeInput`         | Enumerate to specifie type of TextField.       |
| `label`             | text label of TextField.                       |
| `DecorationElement` | input decoration of TextField.                 |
| `onTap`             | callback when you click on TextField .         |
| `hint`              | text hint of textField.                        |
| `errorMsg`          | message to show when TextField isn't validate. |
| `labelStyle`        | style of label TextField                       |
| `errorStyle`        | style of error message TextField               |
| `hintStyle`         | style of hint TextFieldcolor                   |
| `readOnly`          | enable TextField uneditable                    |
| `validator`         | callback validation of TextField               |
| `padding`           | padding of TextField                           |
| `visibility`        | enable visibility of element                   |

### `EmailElement`

> Pre-exsiting element

> check validation of email

> Pre-initialized values

> extends from TextElement


| Properties             | Description                                                         |
| ---------------------- | ------------------------------------------------------------------- |
| `DecorationElement`    | input decoration of TextField.                                      |
| `label`                | text label of TextField.                                            |
| `hint`                 | text hint of textField.                                             |
| `isRequired`           | make textField required in validation                               |
| `errorEmailIsRequired` | error message for textField when it's required                      |
| `errorEmailPattern`    | error message for textField input when it's not email in validation |
| `labelStyle`           | style of label TextField                                            |
| `errorStyle`           | style of error message TextField                                    |
| `hintStyle`            | style of hint TextFieldcolor                                        |
| `readOnly`             | enable TextField uneditable                                         |
| `padding`              | padding of TextField                                                |

### `PasswordElement`


> Pre-exsiting element

> check validation of password

> Pre-initialized values

> show/hide password

> extends from TextElement

| Properties                 | Description                                                                 |
| -------------------------- | --------------------------------------------------------------------------- |
| `DecorationElement`        | input decoration of TextField.                                              |
| `label`                    | text label of TextField.                                                    |
| `hint`                     | text hint of textField.                                                     |
| `errorMsg`                 | message to show when TextField isn't validate.                              |
| `labelStyle`               | style of label TextField                                                    |
| `errorStyle`               | style of error message TextField                                            |
| `hintStyle`                | style of hint TextFieldcolor                                                |
| `readOnly`                 | enable TextField uneditable                                                 |
| `padding`                  | padding of TextField                                                        |
| `enableShowPassword`       | enable eye icon,make password text visible                                  |
| `isRequired`               | make passwordField required                                                 |
| `minLength`                | minimun length accepted by password                                         |
| `hasUppercase`             | make password contains at least one upperCase character                     |
| `hasSpecialCharacter`      | make password contains at least one special character                       |
| `hasDigits`                | make password contains at least one digits                                  |
| `requiredErrorMsg`         | message error to show when password is required                             |
| `minLengthErrorMsg`        | message error to show when password length is less then the specified       |
| `uppercaseErrorMsg`        | message error to show when password doesn't contain any upperCase character |
| `specialCharacterErrorMsg` | message error to show when password doesn't contain any special character   |


### `NumberElement`


> Pre-exsiting element for Number input

> Pre-initialized values

> enabled digitsOnly

> extends from TextElement

| Properties          | Description                                    |
| ------------------- | ---------------------------------------------- |
| `label`             | text label of TextField.                       |
| `hint`              | text hint of textField.                        |
| `DecorationElement` | input decoration of TextField.                 |
| `errorMsg`          | message to show when TextField isn't validate. |
| `labelStyle`        | style of label TextField                       |
| `errorStyle`        | style of error message TextField               |
| `hintStyle`         | style of hint TextFieldcolor                   |
| `readOnly`          | enable TextField uneditable                    |
| `padding`           | padding of TextField                           |
| `isDigits`          | enable only digit number                       |


### `CountryElement`


> Pre-exsiting element for Country input

> Pre-initialized values

> pick country via BottomSheet

> show flag of countries


| Properties              | Description                                   |
| ----------------------- | --------------------------------------------- |
| `DecorationElement`     | input decoration of TextField.                |
| `label`                 | text label of TextField.                      |
| `initValue`             | Initiale Value to country input.              |
| `labelModalSheet`       | Title of modalSheet                           |
| `labelSearchModalSheet` | hint search textfield in BottomSheet          |
| `countryTextResult`     | enumeration get result of selection countries |
| `showFlag`              | show flag of countris in modalsheet           |
| `padding`               | padding of TextField                          |
| `readonly`              | when enable  TextField to be unmodified       |

### `PhoneNumberElement`


> Pre-exsiting element for phone number input

> Pre-initialized values


#### `To Do` in `PhoneNumberElement`

- [ ] pick calling phone via BottomSheet

- [ ] show flag of countries for each calling code

### `Properties`

| Properties          | Description                                                  |
| ------------------- | ------------------------------------------------------------ |
| `DecorationElement` | input decoration of TextField.                               |
| `label`             | text label of TextField.                                     |
| `hint`              | text placeholder for phone number input.                     |
| `initValue`         | initial Value to phone input.                             |
| `errorMsg`          | text error message                                           |
| `validator`         | callback validation of TextField                             |
| `showPrefixFlag`    | enable flag country to be visible at left  of TextField      |
| `showSuffixFlag`    | enable flag country to be visible at rigth of TextField      |
| `padding`           | padding of TextField                                         |
| `showPrefix`        | show calling phone number(get current calling phone of user) |
| `readOnly`          | bool make TextField readOnly                                 |
| `initPrefix`        | (String) initial calling code of the specific country                                |
| `labelModalSheet`   | (String) title of bottom sheet that shown list of calling code of countries                                |

###  `TextAreaElement`

> Pre-exsiting element for multiLine  input (like commentField)
> Pre-initialized values


| Properties     | Description                                                           |
| -------------- | --------------------------------------------------------------------- |
| `maxLines`     | maximum line  to span in textField.                                   |
| `showCounter`  | enable visibility of counterText.                                     |
| `maxCharacter` | The limit on the number of characters that you can type  in textField |

###  `DateElement`

> Pre-exsiting element for date field  input
> Pre-initialized values

| Properties               | Description                                                           |
| ------------------------ | --------------------------------------------------------------------- |
| `id`                     | String,should be unique,
| `initDate`               | (DateTime)  initialize the input field
| `firstDate]`             | (DateTime)  represent earliest allowable Date in date picker
| `lastDate`               | (DateTime)  represent latest allowable Date in date picker
| `format`                 | (DateFormat)  for format the date  that you pick (default :DateFormat.yMd())
| `selectableDayPredicate` | (SelectableDayPredicate)  to enable dates to be selected
| `label`                  | (String) text label of TextField
| `decorationElement`      | input decoration of TextField
| `hint`                   | (String) hint text of textField
| `isRequired`             | (bool) if true,make this field required
| `errorMsg`               | (String) show error message  when the field isn't validate
| `padding`                | (EdgeInsets) padding of textField


###  `RadioGroupElement`

> Pre-exsiting element for radioGroup field  input
> Pre-initialized values

| Properties               | Description                                                           |
| ------------------------ | --------------------------------------------------------------------- |
| `id`                     | String,should be unique,
| `initValue`              | (String)  initialized select value of the radio group  field
| `label`                  | (String) text label of the radio group field
| `labelAlignment`         | (AlignmentGeometry)  label text alignment
| `activeSelectedColor`    | (Color) active select radio in the radio group field
| `decorationElement`      | input decoration of TextField
| `errorMsg`               | (String) error message that will show when error
| `values`                 | (List<String>) list of values of each selected item 
| `valuesLabel`            | (List<String>) list of labels of each item in the radio group
| `padding`                | (EdgeInsets) padding of textField


### `DecorationElement`

> abstract class

> Pre-exsiting inputDecoration for  TextFormField

> Pre-initialized values

> Typically one of `UnderlineDecorationElement` or `OutlineDecorationElement` or `RoundedDecorationElement` can be used.

##### `UnderlineDecorationElement`


| Properties            | Description                                                                       |
| --------------------- | --------------------------------------------------------------------------------- |
| `borderColor`         | The border Color to display when the InputDecorator does not have the focus.      |
| `errorBorderColor`    | The borwidthLineder Color to display when the InputDecorator does have the error. |
| `focusBorderColor`    | The border Color to display when the InputDecorator does  have the focus.         |
| `disabledBorderColor` | The border Color to display when the InputDecorator is disabled.                  |
| `radius`              | radius of the border.                                                             |
| `widthSide`           | The width of this line of the border                                              |
| `filledColor`         | base fill color of the decoration                                                 |
| `focusColor`          | focused fill color of the decoration                                              |


##### `OutlineDecorationElement`

| Properties            | Description                                                                       |
| --------------------- | --------------------------------------------------------------------------------- |
| `borderColor`         | The border Color to display when the InputDecorator does not have the focus.      |
| `errorBorderColor`    | The borwidthLineder Color to display when the InputDecorator does have the error. |
| `focusBorderColor`    | The border Color to display when the InputDecorator does  have the focus.         |
| `disabledBorderColor` | The border Color to display when the InputDecorator is disabled.                  |
| `radius`              | radius of the border.                                                             |
| `widthSide`           | The width of this line of the border                                              |
| `filledColor`         | base fill color of the decoration                                                 |
| `focusColor`          | focused fill color of the decoration                                              |

### `RoundedDecorationElement`

> without BorderSide

| Properties    | Description                          |
| ------------- | ------------------------------------ |
| `radius`      | radius of the border.                |
| `filledColor` | base fill color of the decoration    |
| `focusColor`  | focused fill color of the decoration |


### `ButtonLoginDecorationElement`

> decoration for button login

| Properties              | Description                          |
| ------------------------| ------------------------------------ |
| `shapeButtonLogin`      | shape of the login button.           |
| `backgroundColorButton` | ackground color of the login button  |
| `widthSubmitButton`     | width size of the login button       |
| `elevation`             | elevation of the button              |

* example 
```dart
///decoration Element 
 final decoration = OutlineDecorationElement(
                     filledColor: Colors.white,
                     radius: BorderRadius.only(
                       topLeft: Radius.circular(5.0),
                       topRight: Radius.circular(5.0),
                     ),
                     widthSide: 0.6,
                   );
```

### `PasswordControls`

> define an validation rules for password input

| Properties                 | Description                                                                 |
| ---------------------------| ----------------------------------------------------------------------------|
| `minLength`                | minimum length accepted by password                                         |
| `hasUppercase`             | make password contains at least one upperCase character                     |
| `hasSpecialCharacter`      | make password contains at least one special character                       |
| `hasDigits`                | make password contains at least one digits                                  |



CHANGELOG.md
## [0.23.0]: add RadioGroup field
* create RadioGroupElement
* create Widget `RadioGroupWidget` associated to  RadioGroupElement
* add RadioGroupElement in simpleDynamicForm
## [0.22.0]: add selectChoice field
* create SelectChoiceElement
* create Widget associated to SelectChoiceElement
* add SelectChoiceElement in simpleDynamicForm
## [0.21.0]:
* add new methd `setFieldValueById` in FormController to set Field value 
## [0.20.1+3]:
* add optional validator for passwordElement
* fix default validator for passwordElement
## [0.20.1+2]:
* fix passwordElement
* remove default decoration
## [0.20.1+1]:
* remove default decoration from TextElement
## [0.20.1]:
* apply commonDecoration for missing elements
* improve generate inputBorder
## [0.20.0]
* break changes in LoginForm:
    * all decorationElement and label,hint moved to new Decoration for Login
* create new decoration `DecorationLoginForm`
* add contentPadding to all DecorationElements
* add common decoration in groupElement
* improve PhoneTextElement
## [0.12.3]
* change country url images
## [0.12.2]
* fix rebuild widget and set initValue again
## [0.12.1]   
* remove separator in credit card number text field
## [0.12.0] 
* add dropdown entry mode for card date expiration
* fix bug
## [0.11.0-nullsafety.1] - beta version:
* remove flutter_svg
* replace svg asset with png image
* fix bug
## [0.10.0-nullsafety.1] - beta version:
* remove unnecessary code
## [0.10.0-nullsafety.0] - beta version:
* improve phone text field
* recuperate calling code with phone number
* show country flag by calling code
* country calling code picker
## [0.9.0-nullsafety.1] - beta version:
* fix readme badge
## [0.9.0-nullsafety.0] - beta version:
* migration to null safety feature in dart
* create new form controller for payment form
* add submit button in payment form
* fix bugs in payment form
* remove flag package  that doesn't support null safety
## [0.8.0] - beta version:
* add LoginFormController for LoginForm
* add possibility to show error in the field after validation
* remove loginCallback,textButton,buttonLoginDecorationElement from LoginForm
* change example ,fix Widget test
## [0.7.1] - beta version:
* add submitButton as custom Widget in simpleDynamicForm
## [0.7.0] - beta version:
* remove usage of globalKey (breaking changes)
* create FormController to get data or clear data from form
## [0.6.0] - beta version:
* optimisation replace function generateTextField with statelessWidget
## [0.5.1] - beta version:
* make CVVTextField obscure Text
* create blueprint for CVVTextField
## [0.5.0] - beta version:
* create new form : payment form
* create dateInputField to enter date
* create cardNumberField
* add dateInputElement and cardNumberElement as blueprint for dateInputField and cardNumberField
* fix bugs
## [0.4.5+4] - version alpha:
* fix bug in element validator
## [0.4.5+3] - version alpha:
* fix bug in email element generated
* update dependencies
## [0.4.5+2] - version alpha:
* add password control for loginForm
## [0.4.5+1] - version alpha:
*  remove decorationElement from loginForm
*  add decorationEmailElement and decorationPasswordElement
*  add elevation  and shapeButtonLogin ButtonLoginDecorationElement
*  remove unnecessary code in loginForm
## [0.4.5] - version alpha:
* add DateElement to build date field input
* update docs
## [0.4.4+2] - version alpha:
* add docs
## [0.4.4+1] - version alpha:
* fix bug in clearValues function
## [0.4.4] - version alpha:
* fix bug in some inputs
* add function to clear all values in input fields
* add docs
## [0.4.3] - version alpha:
* add id attribute of element
* recuperate values by id
## [0.4.2+1] - version alpha:
* change internal behavior inside countryTextField
## [0.4.2] - version alpha:
* add visibility property to all element
## [0.4.1+1] - version alpha:
* fix import problem
## [0.4.1] - version alpha:
* add show flag country in phoneText
## [0.4.0+2] - version alpha:
* make MainAxisSize loginForm : MainAxisSize.min
## [0.4.0+1] - version alpha:
* remove MainAxisSize in loginForm
## [0.4.0] - version alpha:
* manage keyboard actions and focusNode,to navigate easy in form
## [0.3.9] - version alpha:
* add textAreaElement for  multiline field
## [0.3.8+1] - version alpha:
* fix dependencies in pubspec
## [0.3.8] - version alpha:
* add PhoneNumberElement
## [0.3.7+6] - version alpha:
* fix readme:example loginForm
## [0.3.7+5] - version alpha:
* add ButtonLoginDecorationElement
* change core of loginForm
* make UnderlineDecoration default decoration
## [0.3.7+4] - version alpha:
* remove deprecated api
## [0.3.7+3] - version alpha:
* fix version in docs
## [0.3.7+2] - version alpha:
* add docs for loginForm
## [0.3.7+1] - version alpha:
* add new form for login: LoginForm
* make passwordError,usernameEmailError
## [0.3.6+1] - version alpha:
* fix readme file
## [0.3.6] - version alpha:
* add decorationElement : UnderlineDecorationElement,OutlineDecorationElement,RoundedDecorationElement
## [0.3.5+2] - version alpha:
* add email validation
* create specifique emailTextField
## [0.3.5+1] - version alpha:
* fix error in readme file
## [0.3.5] - version alpha:
* add validation to passwordTextField
## [0.3.4+2] - version alpha:
* fix error
## [0.3.4+1] - version alpha:
* remove duplicated code
* fix error
## [0.3.4] - version alpha:
* add Country TextField
* add CountryElement
## [0.3.2+1] - version alpha:
* fix error in NumberElement- change validator
## [0.3.2] - version alpha:
## [0.3.1+1] - version alpha:
* fix error
* add numberElement into readme file
## [0.3.1] - version alpha:
* add NumberElement
* create pre-existing number textfield
## [0.3.0] - version alpha:
* add PasswordElement
* make groupelement inside container
* customise groupelement design
## [0.1.1] - version alpha:
 * format files
 * fix code errors
## [0.1.0] - version alpha:
 * generate simple form
 * recuperate all values
 * validate form


example/lib/login_page.dart
import 'package:dynamic_form/dynamic_form.dart';
import 'package:flutter/material.dart';

class LoginPage extends StatefulWidget {
  LoginPage({Key key}) : super(key: key);

  @override
  _LoginPageState createState() => _LoginPageState();
}

class _LoginPageState extends State<LoginPage> {
  final controller = LoginFormController();

  @override
  Widget build(BuildContext context) {
    final decoration = OutlineDecorationElement(
      filledColor: Colors.white,
      radius: BorderRadius.only(
        topLeft: Radius.circular(5.0),
        topRight: Radius.circular(5.0),
      ),
      widthSide: 0.6,
      focusWidthSide: 1.0,
    );
    return LoginForm(
      controller: controller,
      directionGroup: DirectionGroup.Vertical,
      submitLogin: Container(
        width: 250,
        child: ElevatedButton(
          onPressed: () {
            final email = controller.email;
            final password = controller.password;
            controller.validate();
            if (email != "dali") {
              controller.addEmailError("invalid Email not found");
            }
            print("$email,$password");
          },
          child: Text("Log In"),
        ),
      ),
      onlyEmail: false,
      decorationLoginForm: DecorationLoginForm(
        commonDecoration: decoration,
        decorationEmailElement: OutlineDecorationElement(
          contentPadding: EdgeInsets.only(
            left: 12.0,
          ),
          // prefix: Icon(
          //   Icons.person,
          //   size: 20,
          // ),
        ),
        decorationPasswordElement: OutlinePasswordElementDecoration(
            contentPadding: EdgeInsets.only(left: 8),
            prefix: Padding(
              padding: EdgeInsets.all(6.0),
              child: Icon(
                Icons.lock,
                size: 20,
              ),
            ),
            enableVisibilityPassword: false
            // showPasswordWidget: Icon(
            //   Icons.remove_red_eye_outlined,
            // ),
            // hidePasswordWidget: Icon(
            //   Icons.visibility_off_outlined,
            // ),
            ),
        // login: "Username",
        // password: "password",
        hintPassword: "password",
      ),
      paddingFields: const EdgeInsets.all(6),
      passwordControls: PasswordControls.all(
        false,
        passwordMinLength: 3,
      ),
    );
  }
}
